python_modules/jupedsim/jupedsim/distributions.py

```diff
@@ -1,2 +1,440 @@
+        if pt:
+            # Point pt is valid: add it to the samples list and mark it as active
+            samples.append(pt)
+            nsamples += 1
+            active.append(len(samples) - 1)
+            grid.append_point(pt, len(samples) - 1)
+        else:
+            # We had to give up looking for valid points near ref.pt, so remove it
+            # from the list of "active" points.
+            active.remove(idx)
+
+    return samples
+
+
+def get_point(k, refpt, polygon, agent_radius, wall_distance, grid, samples, seed=None, obstacles=None):
+    """Try to find a candidate point relative to refpt to emit in the sample.
+
+    We draw up to k points from the annulus of inner radius r, outer radius 2r
+    around the reference point, refpt. If none of them are suitable (because
+    they're too close to existing points in the sample), return False.
+    Otherwise, return the pt.
+
+    :param k: up to k points are drawn until a reference point is declared as inactive
+    :param refpt: points are drawn around this reference point
+    :param polygon: Polygon in which the points must lie
+    :param agent_radius: minimal distance between points
+    :param wall_distance: minimal distance between points and the polygon
+    :param samples: already placed points
+    :param c_s_l: Cell side length of the grid
+    :param nxny: Number of cells in the x- and y-direction as Tuple: (nx, ny)
+    :param cells: dictionary with key: cell, value: point
+    :param seed: define a seed for random generation
+    :param obstacles: list of polygons, no points can be inside the obstacle
+    """
+    if seed is not None:
+        np.random.seed(seed)
+    box = get_bounding_box(polygon)
+
+    i = 0
+    while i < k:
+        i += 1
+        rho = np.sqrt(np.random.uniform(agent_radius ** 2, 4 * agent_radius ** 2))
+        theta = np.random.uniform(0, 2 * np.pi)
+        pt = refpt[0] + rho * np.cos(theta), refpt[1] + rho * np.sin(theta)
+        if not (box[0][0] <= pt[0] < box[1][0] and box[0][1] <= pt[1] < box[1][1]):
+            # This point falls outside the domain, so try again.
+            continue
+        if point_valid(pt, agent_radius, wall_distance, grid, polygon, samples, obstacles):
+            return pt
+
+    # We failed to find a suitable point in the vicinity of refpt. The Point will be declared as inactive
+    return False
+
+
+def point_valid(pt, agent_radius, wall_distance, grid, polygon, samples, obstacles=None):
+    """ Determines if a point is valid by using a Grid to determine neighbours
+    :param grid: the grid of the polygon
+    :param pt: point that is being checked
+    :param agent_radius:minimal distance between points
+    :param wall_distance: minimal distance between point and the polygon
+    :param polygon: Polygon in which the points must lie
+    :param samples: already placed points
+    :param obstacles: list of polygons, point must not lay within a polygon
```

Comment:
What are obstacles a separate parameter? Polygons can have holes which are our obstacles.

Reply:
should obstacle really be removed as a parameter?
This would result in polygons needing to be a shapely polygon, not just a list of corner points.
To create a shaply polygon with holes you also need these two arguments.

Reply:
done